GooglePlayInstant/LoadingScreen/LoadingScreen.cs

```diff
@@ -0,0 +1,121 @@
+        public LoadingBar LoadingBar;
+
+        public RawImage Background;
+
+        private const int MaxAttemptCount = 3;
+        private AssetBundle _bundle;
+        private int _assetBundleRetrievalAttemptCount;
+
+        private IEnumerator Start()
+        {
+            while (_bundle == null && _assetBundleRetrievalAttemptCount < MaxAttemptCount)
+            {
+                yield return GetAssetBundle(AssetBundleUrl);
+            }
+
+            if (_bundle == null)
+            {
+                Debug.LogError("AssetBundle failed to be downloaded.");
+                yield break;
+            }
+
+            var sceneLoadOperation = SceneManager.LoadSceneAsync(_bundle.GetAllScenePaths()[0]);
+            yield return LoadingBar.FillUntilDone(sceneLoadOperation, LoadingBar.AssetBundleDownloadMaxProportion, 1f);
+        }
+
+        private IEnumerator GetAssetBundle(string assetBundleUrl)
+        {
+            UnityWebRequest webRequest;
+            var downloadOperation = StartAssetBundleDownload(assetBundleUrl, out webRequest);
+
+            yield return LoadingBar.FillUntilDone(downloadOperation,
+                0f, LoadingBar.AssetBundleDownloadMaxProportion);
+
+            if (DidRequestFail(webRequest))
+            {
+                yield return HandleRequestFailed(assetBundleUrl, webRequest);
+            }
+            else
+            {
+                _bundle = DownloadHandlerAssetBundle.GetContent(webRequest);
+            }
+        }
+
+        private IEnumerator HandleRequestFailed(string assetBundleUrl, UnityWebRequest webRequest)
+        {
+            if (_assetBundleRetrievalAttemptCount < MaxAttemptCount)
+            {
+                _assetBundleRetrievalAttemptCount++;
+                Debug.LogFormat("Attempt #{0} at downloading AssetBundle...", _assetBundleRetrievalAttemptCount);
+                yield return new WaitForSeconds(2);
+
+                //TODO: revisit this methodology of setting the loading bar
+                LoadingBar.SetProgress(0f);
+
+                yield return GetAssetBundle(assetBundleUrl);
+            }
+            else
+            {
+                Debug.LogErrorFormat("Error downloading asset bundle: {0}", webRequest.error);
+            }
+        }
+
+        private static bool DidRequestFail(UnityWebRequest webRequest)
+        {
+#if UNITY_2017_1_OR_NEWER
+            var requestFailed = (webRequest.isHttpError || webRequest.isNetworkError);
+#else
+            var requestFailed = webRequest.isError;
+#endif
+            return requestFailed;
+        }
+
+        private static AsyncOperation StartAssetBundleDownload(string assetBundleUrl, out UnityWebRequest webRequest)
+        {
+#if UNITY_2018_1_OR_NEWER
+            webRequest = UnityWebRequestAssetBundle.GetAssetBundle(assetBundleUrl);
+#else
+            webRequest = UnityWebRequest.GetAssetBundle(assetBundleUrl);
+#endif
+
+#if UNITY_2017_2_OR_NEWER
```

Comment:
Same, could consider directly returning. Your call.

Reply:
I have a slight preference towards naming the return value before returning it, at least in this case where the name clarifies what AsyncOperation is being returned.